src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.Data.SqlClientX.Handlers.Connection;
+
+namespace Microsoft.Data.SqlClientX.Handlers.TransportCreation
+{
+    internal class TransportCreationHandler : IHandler<ConnectionRequest>
+    {
+        /// <inheritdoc />
+        public IHandler<ConnectionRequest> NextHandler { get; set; }
+
+        /// <inheritdoc />
+        public async ValueTask Handle(ConnectionRequest request, bool isAsync, CancellationToken ct)
+        {
+            switch (request.TransportCreationParams)
+            {
+                case SharedMemoryTransportCreationParams:
+                    throw new NotImplementedException();
+
+                case NamedPipeTransportCreationParams:
+                    throw new NotImplementedException();
+
+                case TcpTransportCreationParams tcpParams:
+                    request.ConnectionStream = await HandleTcpRequest(tcpParams, ct).ConfigureAwait(false);
+                    break;
+
+                default:
+                    throw new InvalidOperationException("");
+            }
+
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(request, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        private ValueTask<Stream> HandleNamedPipeRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private ValueTask<Stream> HandleSharedMemoryRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private async Task<Stream> HandleTcpRequest(TcpTransportCreationParams request, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
```

Comment:
I'd recommend putting this in the IP address loop too - the `Socket.Poll` call can take a while, and in the absence of a CancellationToken overload I think it's worth giving the caller the opportunity to exit the loop early.

Reply:
Sounds good to me 👍